goat_storytelling_agent/storytelling_agent.py

```diff
@@ -50,366 +65,484 @@ def _query_chat(messages, retries=3, request_timeout=120):
-
-        act['chapter_scenes'] = {}
-        chapters = [text.strip() for text in act_scenes[:]
-                    if (text and text.strip())]
-        current_ch = None
-        merged_chapters = {}
-        for snippet in chapters:
-            if snippet.isnumeric():
-                ch_num = int(snippet)
-                if ch_num != current_ch:
-                    current_ch = snippet
-                    merged_chapters[ch_num] = ''
-                continue
-            if merged_chapters:
-                merged_chapters[ch_num] += snippet
-        ch_nums = list(merged_chapters.keys()) if len(
-            merged_chapters) <= len(act_chapters[i]) else act_chapters[i]
-        merged_chapters = {ch_num: merged_chapters[ch_num]
-                           for ch_num in ch_nums}
-        for ch_num, chapter in merged_chapters.items():
-            scenes = re.split(r'Scene \d+.{0,10}?:', chapter)
-            scenes = [text.strip() for text in scenes[1:]
-                      if (text and (len(text.split()) > 3))]
-            if not scenes:
-                continue
-            act['chapter_scenes'][ch_num] = scenes
-    return all_messages, plan
-
-
-def prepare_scene_text(text):
-    lines = text.split('\n')
-    ch_ids = [i for i in range(5)
-              if 'Chapter ' in lines[i]]
-    if ch_ids:
-        lines = lines[ch_ids[-1]+1:]
-    sc_ids = [i for i in range(5)
-              if 'Scene ' in lines[i]]
-    if sc_ids:
-        lines = lines[sc_ids[-1]+1:]
-
-    placeholder_i = None
-    for i in range(len(lines)):
-        if lines[i].startswith('Chapter ') or lines[i].startswith('Scene '):
-            placeholder_i = i
+def _query_chat_llamacpp(endpoint, messages, retries=3, request_timeout=120, max_tokens=4096, extra_options={}):
+    headers = {'Content-Type': 'application/json'}
+    prompt = ''.join(generate_prompt_parts(messages))
+    print(f"\n\n========== Submitting prompt: >>\n{prompt}", end="")
+    sys.stdout.flush()
+    response = requests.post(
+        f"{endpoint}tokenize", headers=headers, data=json.dumps({"content": prompt}),
+        timeout=request_timeout, stream=False)
+    tokens = [1, *response.json()["tokens"]]
+    data = {
+        "prompt": tokens,
+        "stream": True,
+        "n_predict": max_tokens - len(tokens),
+        **extra_options,
+    }
+    jdata = json.dumps(data)
+    request_kwargs = dict(headers=headers, data=jdata, timeout=request_timeout, stream=True)
+    response = requests.post(f"{endpoint}completion", **request_kwargs)
+    result = bytearray()
+    if messages and messages[-1]["role"] == "assistant":
+        result += messages[-1]["content"].encode("utf-8")
+    is_first = True
+    for line in response.iter_lines():
+        line = line.strip()
+        if not line:
+            continue
+        if line.startswith(b"error:"):
+            retries -= 1
+            print("\nError(retry={retries}): {line!r}")
```

Comment:
missing `f` for the formatting to trigger.

Reply:
I have to give you credit for that one. :)

You might want to make your own pull request vs master with these fixes since this one already got merged.